option_simple_close (#1096)
lightning/bolts: 02-peer-protocol.md

1. one side indicates it wants to clear the channel (and thus will accept no new HTLCs)
2. once all HTLCs are resolved, the final channel close negotiation begins.

>         +-------+                              +-------+

Reviewer: IMO we should have awareness of this be signalled via feature bit rather than remove the old version. Then in the future, we can start to make the bit required to enforce the switch over on the protocol level.

        |       | <complete all pending HTLCs> |       |
        |   A   |                 ...          |   B   |
        |       |                              |       |
        |       |--(3a)- closing_complete Fee->|       |
        |       |<-(3b)- closing_complete Fee--|       |
        |       |<-(4a)- closing_sig ----------|       |

>         |       |--(4b)- closing_sig --------->|       |

Reviewer 2: This doesn't seem like the typical use case where one side initiates the close and the other signs.  Some explanation somewhere is needed:

- Is the idea here that both sides can contribute to the closing fee if they want?
- Does B's `closing_complete` override A's?  And in that case, why does B send `closing_sig`?
- Did both A and B send `closing_complete` at the same time, so we have 2 flows happening at once?

Reviewer 3: There is one flow happening in each direction, where the sender of `closing_complete` is paying the fee.
Let me know if the proposed diagram [here](https://github.com/lightning/bolts/pull/1096#issuecomment-1743283241) makes it easier to reason about.

Reviewer 2: > There is one flow happening in each direction, where the sender of `closing_complete` is paying the fee. 

Do we really expect dual flows to happen like that in practice?  If the other party offers to pay for the closing transaction, why would we want to send `closing_complete` at all?

> Let me know if the proposed diagram [here](https://github.com/lightning/bolts/pull/1096#issuecomment-1743283241) makes it easier to reason about.

Your diagram helps to clarify corner cases but probably still needs some high-level explanation in the spec.

Reviewer 3: > Do we really expect dual flows to happen like that in practice?

I think we do, because each side should independently decide what fee they'd like to pay and immediately send `closing_complete` (at least that's how I implemented it). Since the channel has already stopped relaying HTLCs at that point, both peers are likely eager to get their funds back asap to reinvest them into another channel.

  - MAY send a `shutdown` before a `channel_ready`, i.e. before the funding transaction has reached `minimum_depth`.
  - if there are updates pending on the receiving node's commitment transaction:
    - MUST NOT send a `shutdown`.

>   - MUST NOT send multiple `shutdown` messages.

Reviewer: Should't this instead be gated on awareness/negotiation of the new feature bit? Old nodes will still reject multiple shutdowns, so it should be feature gated.

Author: No, everyone always accepted multiple shutdown messages, because we *required* that they be retransmitted.  This requirement was always ignored!

    3. if (and only if) `option_shutdown_anysegwit` is negotiated:
      * `OP_1` through `OP_16` inclusive, followed by a single push of 2 to 40 bytes
        (witness program versions 1 through 16)
    4. if (and only if) `option_simple_close` is negotiated:
      * `OP_RETURN` followed by one of:
        * `6` to `75` inclusive followed by exactly that many bytes

>         * `76` followed by `76` to `80` followed by exactly that many bytes

Author: Script semantics: 
1. 6-75 push this many bytes on the stack.
2. 76 (aka OP_PUSHDATA1): next byte says how many bytes to push on the stack

>         * `76` followed by `76` to `80` followed by exactly that many bytes

Author: Added that to Rationale.

>         * `76` followed by `76` to `80` followed by exactly that many bytes

Reviewer 3: Thanks! Bitcoin script is weird :sweat_smile:

    - SHOULD send a `closing_complete` message.

The sender of `closing_complete` (aka. "the closer"):
  - MUST set `fee_satoshis` to a fee less than or equal to its outstanding balance, rounded down to whole satoshis.
  - MUST set `has_closee_output` to 0 if it would be dust (< 540 satoshis for P2SH, < 330 satoshis for P2WSH and P2TR)
  - SHOULD set `has_closer_output` to 0 if it considers its own remaining balance to be uneconomic.
  - Otherwise MUST set `has_closer_output` to 1.
  - If it sets `has_closer_output` to 1:
    - MUST set `signature_with_closee_output` to a valid signature of a transaction with both closer and closee outputs.
	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only a closer output.
  - Otherwise: (`has_closer_output` is 0):
    - MUST set `signature_with_closee_output` to a valid signature of a transaction with only the closee output.
	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output as described in [BOLT 3](03-transactions.md#closing-transaction).

The receiver of `closing_complete` (aka. "the closee"):

>   - If `has_closee_output` is 1 and the closer's output would be dust (< 540 satoshis for P2SH, < 330 satoshis for P2WSH and P2TR):

Reviewer 2: ```suggestion
  - If `has_closer_output` is 1 and the closer's output would be dust (< 540 satoshis for P2SH, < 330 satoshis for P2WSH and P2TR):
```

    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - if either `signature_with_closee_output` or `signature_without_closee_output` is not valid for the closing transactions specified in [BOLT #3](03-transactions.md#closing-transaction) OR non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - Otherwise:
    - MUST select one of the signatures (and thus, transactions) to respond to.
	- MUST sign and broadcast that transaction.
    - MUST send `closing_sig`.

The sender of `closing_sig`:
  - if it selected `signature_with_closee_output` to broadcast:
    - MUST set `has_closee_output` to 1.
	- MUST set `signature` to a valid signature on that transaction.
  - otherwise (it selected `signature_without_closee_output`):
    - MUST set `has_closee_output` to 0.
	- MUST set `signature` to a valid signature on that transaction.

The receiver of `closing_sig`:

>   - MUST broadcast the transaction indicated by `has_closee_output`.

Reviewer 2: No signature verification before broadcast?

### Rationale

The close protocol is designed to avoid any failure scenarios caused by fee disagreement, since each side offers to pay its own desired fee.

Multiple signatures are given, so each side can choose whether it wants to include its own output.  In the case of large fees and tiny channels, where neither side wants its output, the use of an OP_RETURN simply cleans up the dangling unspent transaction output.  This is expected to be extremely rare.

> Note that there is usually no reason to pay a high fee for rapid processing, since an urgent child could pay the fee on the closing transactions' behalf.

Reviewer 2: We could also mention RBF of the closing transaction here.

Reviewer 4: Also I fail to understand how we could ever have a transaction be urgent enough to push for confirmation if it involved burning all of the outputs in an OP_RETURN.

  - MUST set `fee_satoshis` so that at least one output is not dust.
  - MUST use the last send and received `shutdown` `scriptpubkey` to generate the closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction).
  - If it sets `signature` fields, MUST set them as valid signature using its `funding_pubkey` of:
    - `closer_noclosee`: closing transaction with only the local ("closer") output.
    - `no_closer_closee`: closing transaction with only the remote ("closee") output.
    - `closer_and_closee`: closing transaction with both the closer and closee outputs.
  - If the local outstanding balance (in millisatoshi) is less than the remote outstanding balance:
    - MUST NOT set `closer_no_closee`.
    - MUST set exactly one of `no_closer_closee` or `closer_and_closee`.
    - MUST set `no_closer_closee` if the local output amount is dust.
    - MAY set `no_closer_closee` if it considers the closee output amount uneconomic AND its `scriptpubkey` is not `OP_RETURN`.
  - Otherwise (not lesser amount, cannot remove own output):
    - MUST NOT set `no_closer_closee`.

>     - MUST set both `closer_no_closee` and `closer_and_closee`.

Reviewer 3: Why require setting both here? If the closee's output is below dust, we should be allowed to only set `closer_no_closee`, because `closer_and_closee` would lead to a non-standard transaction?

I understand that when the closee's output is *not* below dust, we must still set `close_no_closee` to allow them to omit their output though.

So I think the requirements for that section should be:

```
 - Otherwise (not lesser amount, cannot remove own output):
    - MUST NOT set `no_closer_closee`.
    - If the closee's output amount is dust:
      - MUST set `closer_no_closee`.
      - SHOULD NOT set `closer_and_closee`.
    - Otherwise:
      - MUST set both `closer_no_closee` and `closer_and_closee`.
```

The receiver of `closing_complete` (aka. "the closee"):
  - If `fee_satoshis` is greater than the closer's outstanding balance:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - Select a signature for validation:
    - if the local output amount is dust:
      - MUST use `closer_no_closee`.
    - otherwise, if it considers the closee output amount uneconomic AND its `scriptpubkey` is not `OP_RETURN`:
      - MUST use `closer_no_closee`.
    - otherwise, if `closer_and_closee` is present:
      - MUST use `closer_and_closee`.
    - otherwise:
      - MUST use `no_closer_closee`.
  - If the selected signature field does not exist:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - If the signature field is not valid for the corresponding closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction):

>     - MUST ignore `closing_complete`.

Reviewer 3: I think I see, what you have in mind is that each side only remembers the last remote `shutdown` they received and the last `shutdown` they sent (and the corresponding scripts), so if there is a race between `shutdown` from one side and `closing_complete` from the other, there will be a mismatch in the scripts used.

It is important to highlight that this will work and converge because when sending `shutdown`, we expect the remote to answer with their `shutdown` (even if they haven't changed their script) before sending our `closing_complete`. Am I correct?

>     - MUST ignore `closing_complete`.

Author: Yes, I don't know if you received my last shutdown.  But as you say, it will work out, because you'll re-xmit closing_complete when you do.

  - If the signature field is non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - MUST sign and broadcast the corresponding closing transaction.
  - MUST send `closing_sig` with a single valid signature in the same tlv field as the `closing_complete`.

The receiver of `closing_sig`:
  - if `tlvs` does not contain exactly one signature:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - if `tlvs` does not contain one of the tlv fields sent in `closing_complete`:
    - MUST ignore `closing_complete`.
  - if the signature field is not valid for the corresponding closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction):
    - MUST ignore `closing_complete`.

>   - of the signature field is non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:

Reviewer 3: nit:
```suggestion
  - if the signature field is non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
```

  - If the selected signature field does not exist:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - If the signature field is not valid for the corresponding closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction):
    - MUST ignore `closing_complete`.
  - If the signature field is non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - MUST sign and broadcast the corresponding closing transaction.
  - MUST send `closing_sig` with a single valid signature in the same tlv field as the `closing_complete`.

The receiver of `closing_sig`:
  - if `tlvs` does not contain exactly one signature:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - if `tlvs` does not contain one of the tlv fields sent in `closing_complete`:
    - MUST ignore `closing_sig`.
  - if the signature field is not valid for the corresponding closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction):

>     - MUST ignore `closing_complete`.

Reviewer 3: ```suggestion
    - MUST ignore `closing_sig`.
```

No funds are at risk when that happens, but the channel must be force-closed as
the closing transaction will likely never reach miners.

### Closing Negotiation: `closing_complete` and `closing_sig`

Once shutdown is complete, the channel is empty of HTLCs, there are no commitments
for which a revocation is owed, and all updates are included on both commitments,
the final current commitment transactions will have no HTLCs.

> Each peer says what fee it will pay, and the other side simply signs that transaction.  The only complexity comes from allowing each side to omit its own output should it be uneconomic.

Reviewer 5: small nit: we can split the line here in two

> This process will be repeated every time a `shutdown` message is received, which allows re-negotiation.

Reviewer 3: Right now redundant `shutdown` messages are ignored: it won't be the case anymore (especially with taproot since we'll need to use the nonces from that message). Should we now allow the sender to use a different closing `scriptpubkey` than the one they used in previous `shutdown`? Or should we keep restricting it for simplicity?

Reviewer 5: I guess allowing different `scriptpubkey` should be possible and the closer want to redirect the funds in another `scriptpubkey`.

ofc at this point we should also send it back inside the `closing_complete` msg, imho

Author: Yes, let them change it. Why not?

Reviewer 5: I guess we should have this because we can have concurrency around the network? and getting the confirmation on what scriptpubkey we close the channel can be useful to avoid unnecessary mess,  right?

Author: No concurrency, reconnection requirements already say you re-xmit on every reconnect:

```
  - upon reconnection:
    - if it has sent a previous `shutdown`:
      - MUST retransmit `shutdown`.
```

Reviewer 5: Ah I was missing it! good point, but in this case, the concurrency can happen if I send two subsequent shutdown messages with different `scriptpubkey`. I do not know why people should do it, but the spec allows it

1. type: 40 (`closing_complete`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u64`:`fee_satoshis`]
   * [`u8`: `has_closer_output`]

>    * [`signature`:`signature_with_closee_output`]

Author: The existing dust trimming logic is static, and applies to several different cases.  At this point, you know exactly what the spend cost is, so let's be explicit.

   * [`signature`:`signature_without_closee_output`]

1. type: 41 (`closing_sig`)
2. data:
   * [`channel_id`:`channel_id`]

>    * [`u8`: `closee_output`]

Author: I think we probably will, and make taproot dependent on this feature so it's easy...

(I haven't actually implemented this to see if there are any gotchas, which is needed anyway)

   * [`signature`:`signature`]

#### Requirements

Note: the details and requirements for the transaction being signed are in [BOLT 3](03-transactions.md#closing-transaction)).

Both nodes:
  - After a `shutdown` has been received, AND no HTLCs remain in either commitment transaction:
    - SHOULD send a `closing_complete` message.

The sender of `closing_complete` (aka. "the closer"):
  - MUST set `fee_satoshis` to a fee less than or equal to its outstanding balance, rounded down to whole satoshis.
  - SHOULD set `has_closer_output` to 0 if it considers its own remaining balance to be uneconomic.
  - Otherwise MUST set `has_closer_output` to 1.
  - If it sets `has_closer_output` to 1:
    - MUST set `signature_with_closee_output` to a valid signature of a transaction with both closer and closee outputs.
	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only a closer output.
  - Otherwise: (`has_closer_output` is 0):
    - MUST set `signature_with_closee_output` to a valid signature of a transaction with only the closee output.

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Reviewer 6: Might be missing something but I don't think standardness will allow this case into mempool/relay.

1 native segwit input with 1 OP_RETURN output is ~61 bytes of non-witness data.

Bitcoin Core would allow 65+ starting with Bitcoin Core 25.0: https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-25.0.md#notable-changes

but sub-64 is still not allowed.

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Author: It's defined in BOLT 3 which covers the actual tx construction...

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Reviewer 7: If you're in a cooperative close scenario, but you've decided you're happy to abandon all your channel funds, why not donate them to your counterparty rather than miners? ie, add what would have been your balance to your counterparty's output value. If your counterparty then wants to move some of that to fees, they can re-negotiate the cooperative close and do an rbf?

Adding null output handling doesn't seem in line with the goal of making things simpler and less special case-y.

If you are signing a "send all my funds to fees" tx, you should probably be doing it in an aggregatable way (`ANYONECANPAY|NONE` or `ANYONECANPAY|ALL`) so that in the event of a bunch of channels being closed because block space is scarce and expensive, the 10+ byte outputs of all those channels can be merged.

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Reviewer 3: But why add any of this complexity in the first place? If we're in that scenario, that means the balance of the closer was very low (otherwise they wouldn't omit their output) which means that the balance on the closee side is large enough that they'll never want to omit it (it doesn't make sense to allow channels with a total capacity that is ~dust).

We should only send a signature for a transaction that contain the closee output, and omit the case where the transaction doesn't contain any real output, don't we?

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Reviewer 6: > why not donate them to your counterparty rather than miners?

Adding financial incentives to goad coop closes seems suboptimal.

> We should only send a signature for a transaction that contain the closee output, and omit the case where the transaction doesn't contain any real output, don't we?

The non-initiator can burn their dust later if they so choose, and in any manner they choose, without having to specify exactly how. I think that's a vote in favor of it.

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Reviewer 7: > Adding additional financial incentives to goad coop closes seems suboptimal.

I think the logic here should be the same as for dust: "if this output will never be worth spending, don't create it". If done that way, there's no opportunity for goading -- first, you should never allow your balance to drop to that level, and in the event that it somehow does (your channel was down while fees were rising, you changed your configuration and increased your dust value), you simply unconditionally abandon the channel when it's closed.

If your threshold for a spendable output is X, you estimate a cooperative close fee as F, you estimate a unilateral close fee as U (U>F), and your balance is B, then the logic is:

 * if `B > X`, reject any forward that would drop B below X. (or, for balance privacy, if `B < X + htlc_maximum_msat`, reject any forward to this channel)
 * if `B < X`, always give the balance to your counterparty.
 * if `B < X+U` consider rebalancing to increase B above X+U (or, if supported, rebalancing(draining) the channel to 0 and closing it).
 * if `0 < B < X+F`, don't automatically close the channel. if the user insists anyway, give the balance to your counterparty.
 * if `X+F < B < X+U` and rebalancing isn't possible, consider proposing a mutual close.

If you initiate a mutual close in the last case, you get an output with (B-F > X) funds, and can choose a low fee rate that will be slow to confirm (it's only a small output anyway, so you probably don't care very much when it's confirmed), leaving your counterparty to feebump if they want access to their funds sooner.

(Also, burning funds to fees doesn't remove the incentive to game a channel close, it at best moves the incentive from your counterparty to miners, and, if your counterparty was going to pay lots of fees anyway, just funds your counterparty less directly)

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Author: Two points:

- One day, splice-to-close will be a thing, and you'll get to carefully craft your close tx to do arbitrary things.
- I don't expect to ever see a "no output" close on mainnet, but we have to cover it.

> 	- MUST set `signature_without_closee_output` to a valid signature of a transaction with only the null output.

Reviewer 7: > I don't expect to ever see a "no output" close on mainnet, but we have to cover it.

mainnet is the only place where fees might plausibly render an entire channel to be dust making burning the entire balance to fees sensible. If the feature's not useful there, it isn't useful anywhere.

In the event that a pair of nodes really do want to abandon a channel balance, they can manually specify a scriptPubKey of  "OP_RETURN [padding]" or "p2pkh[G]" to either permanently burn the funds (ultrasound money!) or donate to some future miner that's got some free space in a block. Anyway, if you want to make your simple close unnecessarily complicated, I guess that's your call, so I won't harp on this further...

  - otherwise (it selected `signature_without_closee_output`):
    - MUST set `closee_output` to 0.
	- MUST set `signature` to a valid signature on that transaction.

The receiver of `closing_sig`:
  - SHOULD broadcast the transaction indicated by `closee_output`.

### Rationale

The close protocol is designed to avoid any failure scenarios caused by fee disagreement, since each side offers to pay its own desired fee.

Multiple signatures are given, so each side can choose whether it wants to include its own output.  In the case of large fees and tiny channels, where neither side wants its output, the use of an OP_RETURN simply cleans up the dangling unspent transaction output.  This is expected to be extremely rare.

Note that there is usually no reason to pay a high fee for rapid processing, since an urgent child could pay the fee on the closing transactions' behalf.

If the closer proposes a transaction which will not relay (either its own output is dust, or the fee rate it proposes is too low), it doesn't harm the closee to sign the transaction.

Similarly, if the closer proposes a high fee, it doesn't harm the closee to sign the transaction, as the closer is paying.

> There's a slight game where each side would prefer the other side pay the fee, and proposes a minimal fee.  If neither side proposes a fee which will relay, the negotiation can occur again, or the final commitment transaction can be spent.  In practice, the opener has an incentive to offer a reasonable closing fee, as they would pay the fee for the commitment transaction, which also costs more to spend.

Reviewer 8: This discussion of the game theory hints at an issue I've always had with the "opener pays" rule. I do understand the pathological scenario that the rule is meant to avoid (where a malicious node can drain the funds of a victim node by repeatedly opening and abandoning channels to it), but this new proposal creates a perverse incentive.

Olav opens a channel to Kristján. Some time later, Kristján wants the channel closed. Under this new proposal, Kristján's fee-optimal channel-closing strategy would be to transmit a bogus `error` message to Olav so that Olav will be forced (by protocol convention) to unilaterally close the channel. Thusly, Kristján would obtain an immediately spendable output and would avoid paying any fees.

Author: That's always true.

But K can also mutual close at <1 sat per byte.

Reviewer 8: How can any mutual closing transaction use <1 sat/vB fee rate? The transaction would not propagate.

And you're right that it's always true. It has always irked me. I swear I have seen some cases where nodes to which I have opened channels have sent me `error` messages for no reason other than to get me to unilaterally close. I think it was a design mistake to have commitment transactions reward the non-closing party with an immediately spendable output. I think commitment transactions should time-lock _both_ outputs to discourage baiting a peer into unilaterally closing.

Reviewer 3: That will disappear once we have package relay, because at that point commitment transactions will pay no fees. The fees will have to be paid using CPFP on an ephemeral anchor. Also, if you send bogus `error`s, nodes will start blacklisting you at some point and won't accept new channels from you: this isn't the case today, but will definitely be in the future.

Reviewer 9: The opener should at least be on the hook for whatever the commitment fee is. Otherwise, as the non-opener, it's cheaper to FC and wait out the time lock. Also, if there's nothing on the non-opener side, is FC the only options since there's nothing on that side to pay the fee.

Reviewer 4: I want to challenge this idea of "it's cheaper to FC and wait out the time lock". Having money locked up carries opportunity cost of the entire output. Yes you may save in fees, but your outputs are tied up in chain resolution and that is a cost in its own right.

> Also, if there's nothing on the non-opener side, is FC the only options since there's nothing on that side to pay the fee.

Why would a node be concerned about aggressively closing a channel that has no capital tied up in it?

Reviewer 9: Current coop close is ~168 vB (2 x `P2WPKH` outputs) to ~192 vB (2 x `P2TR` outputs). Sweeping the timelock is 121-123 vB.. so ~35-50% smaller tx. I agree with the time cost, but if you are closing due to months of inactivity, what's another up to two weeks. But the opener would be doubly screwed since they would pay the FC (+ anchors) and then have to pay to sweep. I supposed the opener can "deter" this by keeping max commit fees low.

Closing channels should be discouraged, but having dead/inactive channel is worse.

In my mind, game theory would have me coop close at a very low fee if the balance is mostly on the other side or FC if it's mostly on my side and I'm not the opener and commit fee is high enough. 

>Why would a node be concerned about aggressively closing a channel that has no capital tied up in it?

More of a hypothetical.. how does a closer pay if there's little to nothing on their side.

I agree a change to channel closing is needed, but 100% opener to 100% closer is a pretty drastic change. I would like to see the opener using (some of) the commit fee already reserved and/or splitting the fee based on balance.

No funds are at risk when that happens, but the channel must be force-closed as
the closing transaction will likely never reach miners.

`OP_RETURN` is only standard if followed by PUSH opcodes, and the total script is 83 bytes or less.  We are slightly stricter, to only allow a single PUSH, but there are two forms in script: one which pushes up to 75 bytes, and a longer one (OP_PUSHDATA1) which is needed for 76-80 bytes.

### Closing Negotiation: `closing_complete` and `closing_sig`

Once shutdown is complete, the channel is empty of HTLCs, there are no commitments
for which a revocation is owed, and all updates are included on both commitments,
the final current commitment transactions will have no HTLCs.

Each peer says what fee it will pay, and the other side simply signs that transaction.  The lesser-paid peer (if either is) can opt to omit their own output from the closing tx.

This process will be repeated every time a `shutdown` message is received, which allows re-negotiation (and RBF).

1. type: 40 (`closing_complete`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u64`:`fee_satoshis`]

>    * [`u32`:`sequence`]

Reviewer 3: What about also including `lockTime` like we do in `open_channel2`?

  - If it sets `signature` fields, MUST set them as valid signature using its `funding_pubkey` of:
    - `closer_no_closee`: closing transaction with only the local ("closer") output.
    - `no_closer_closee`: closing transaction with only the remote ("closee") output.
    - `closer_and_closee`: closing transaction with both the closer and closee outputs.
  - If the local outstanding balance (in millisatoshi) is less than the remote outstanding balance:
    - MUST NOT set `closer_no_closee`.
    - MUST set exactly one of `no_closer_closee` or `closer_and_closee`.
    - MUST set `no_closer_closee` if the local output amount is dust.
    - MAY set `no_closer_closee` if it considers the local output amount uneconomic AND its `scriptpubkey` is not `OP_RETURN`.
  - Otherwise (not lesser amount, cannot remove own output):
    - MUST NOT set `no_closer_closee`.
    - If the closee's output amount is dust:
      - MUST set `closer_no_closee`.
      - SHOULD NOT set `closer_and_closee`.
    - Otherwise:
      - MUST set both `closer_no_closee` and `closer_and_closee`.

> The receiver of `closing_complete` (aka. "the closee"):
>   - If `fee_satoshis` is greater than the closer's outstanding balance:
>     - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.

Reviewer 4: Worth noting here that this introduces an edge case *if* we decide to rip out the original coop-close protocol. This would make it such that if you needed to coop-close the channel as the target of a channel-open that you would have no choice but to force close it. I think that everyone loses in this scenario.

```suggestion
The receiver of `closing_complete` (aka. "the closee"):
  - If `fee_satoshis` is greater than the closer's outstanding balance:
    - MAY either send a `warning` and close the connection, or send an `error` and fail the channel.
    - MAY choose to cover the deficit from their own output.
```

This way, if the closer tries to grief you by suggesting an insane fee, you can reject it, but otherwise if it's a fee you'd be willing to pay and it beats your fee liability if you have to force close, you come out ahead. Note that the only realistic scenario where this happens is if you just had a channel opened to you, received no inbound satsflow on the channel but you've decided to close the channel for whatever reason. In this case you may still want to be a good actor and coop-close the channel, but this construction (along with the removal of the existing coop-close) would force you to be more aggressive.

Reviewer 4: Withdrawn. I misunderstood some of the other aspects of this protocol.

  - If `sequence` is equal to 0xFFFFFFFF:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - Select a signature for validation:
    - if the local output amount is dust:
      - MUST use `closer_no_closee`.
    - otherwise, if it considers the closee output amount uneconomic AND its `scriptpubkey` is not `OP_RETURN`:
      - MUST use `closer_no_closee`.
    - otherwise, if `closer_and_closee` is present:
      - MUST use `closer_and_closee`.
    - otherwise:

>       - MUST use `no_closer_closee`.

Reviewer 3: That is the gentleman's approach, but that isn't the most efficient one for the closee. If the closee receives a proposal that contains `no_closer_closee`, that is the option that pays the most fees, so the closee has an incentive to choose this one instead of `closer_and_closee`?

In practice though, if the sender sends `no_closer_closee`, they won't send `closer_and_closee`, so those cases are only theoretical?

  - if the signature field is not valid for the corresponding closing transaction specified in [BOLT #3](03-transactions.md#closing-transaction):
    - MUST ignore `closing_complete`.
  - if the signature field is non-compliant with LOW-S-standard rule<sup>[LOWS](https://github.com/bitcoin/bitcoin/pull/6769)</sup>:
    - MUST either send a `warning` and close the connection, or send an `error` and fail the channel.
  - otherwise:
    - MUST sign and broadcast the corrsponding closing transaction.

### Rationale

The close protocol is designed to avoid any failure scenarios caused by fee disagreement, since each side offers to pay its own desired fee.

If one side has less funds than the other, it may choose to omit its own output, and in this case dust MUST be omitted, to ensure the resulting transaction can be broadcast.

The corner case where fees are so high that both outputs are dust is addressed in two ways: paying a low fee to avoid the problem, or using an OP_RETURN (which is never "dust").

Note that there is usually no reason to pay a high fee for rapid processing, since an urgent child could pay the fee on the closing transactions' behalf.

> However, sending a new `shutdown` message overrides previous ones, so you can negotiate again (even changing the output address) if you want: in this case there's a race where you could receive a `closing_complete` for the previous output address, and the signature won't validate.  In this case, ignoring the `closing_complete` is the correct behaviour, as the new `shutdown` will trigger a new `closing_complete` with the correct signature.  This assumption that we only remember the last-sent of any message is why so many cases of bad signatures are simply ignored.

Reviewer 10: `so you can negotiate again (even changing the output address) if you want`

doesn't this override the safety feature of upfront shutdown?

Reviewer 3: This is of course if you don't have `upfront_shutdown` activated, we can make that clearer.
